File: app/src/main/java/ui/fragment/BaseFragment.java
package ui.fragment;

import android.os.Bundle;

import androidx.annotation.IdRes;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;

import ui.activity.MainActivity;
import utils.KeyboardInputHelper;
import viewmodel.CategoryViewModel;
import viewmodel.TodoViewModel;


public class BaseFragment extends Fragment {

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
    }

    protected void hideKeyboard() {
        if (getContext() == null)
            return;
        KeyboardInputHelper.getKeyboardInput().hideKeyboard(getContext());
    }

    protected void showKeyboard() {
        if (getContext() == null)
            return;
        KeyboardInputHelper.getKeyboardInput().showKeyboard(getContext());
    }

    protected void back() {
        hideKeyboard();
        getParentFragmentManager().popBackStack();
    }

    protected void back(String fragmentTagName) {
        hideKeyboard();
        getParentFragmentManager().popBackStack(fragmentTagName, FragmentManager.POP_BACK_STACK_INCLUSIVE);
    }

    protected void back(@IdRes int fragmentID) {
        hideKeyboard();
        getParentFragmentManager().popBackStack(fragmentID, FragmentManager.POP_BACK_STACK_INCLUSIVE);
    }

    protected FragmentTransaction transit(FragmentTransaction transaction, int transitionType) {
        return transaction.setTransition(transitionType);
    }

    protected FragmentTransaction transit(FragmentTransaction transaction) {
        return transaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN);
    }

    protected TodoViewModel getTodoViewModel() {
        if (getActivity() == null)
            return null;

        return ((MainActivity) getActivity()).getTodoViewModel();
    }

    protected CategoryViewModel getCategoryViewModel() {
        if (getActivity() == null)
            return null;

        return ((MainActivity) getActivity()).getCategoryViewModel();
    }

}

File: app/src/main/java/ui/fragment/BaseFragment.kt
package ui.fragment

import android.os.Bundle
import androidx.annotation.IdRes
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentTransaction
import androidx.fragment.app.activityViewModels
import hlv.cute.todo.App
import utils.AppResourcesProvider
import utils.KeyboardUtil
import viewmodel.CategoryViewModel
import viewmodel.SearchViewModel
import viewmodel.TodoViewModel

open class BaseFragment : Fragment() {

    protected val todoViewModel by activityViewModels<TodoViewModel>()
    protected val categoryViewModel by activityViewModels<CategoryViewModel>()
    protected val searchViewModel by activityViewModels<SearchViewModel>()

    protected val provideResource = AppResourcesProvider(App.get()?.appContext!!)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        hideKeyboard()
    }

    private fun hideKeyboard() {
        context?.let {
            KeyboardUtil.hideKeyboard(it)
        }
    }

    protected fun showKeyboard() {
        KeyboardUtil.showKeyboard(context!!)
    }

    protected fun back() {
        hideKeyboard()
        parentFragmentManager.popBackStack()
    }

    protected fun back(fragmentTagName: String?) {
        hideKeyboard()
        parentFragmentManager.popBackStack(
            fragmentTagName,
            FragmentManager.POP_BACK_STACK_INCLUSIVE
        )
    }

    protected fun back(@IdRes fragmentID: Int) {
        hideKeyboard()
        parentFragmentManager.popBackStack(fragmentID, FragmentManager.POP_BACK_STACK_INCLUSIVE)
    }

    protected fun transit(
        transaction: FragmentTransaction,
        transitionType: Int
    ): FragmentTransaction {
        return transaction.setTransition(transitionType)
    }

    protected fun transit(transaction: FragmentTransaction): FragmentTransaction {
        return transaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
    }

    override fun onDestroyView() {
        super.onDestroyView()
        hideKeyboard()
    }
}